### PR TITLE
ENH: `scipy.stats.qmc.Sobol`: allow 32 or 64 bit computation

### PR DESCRIPTION
**Pull request by tupui**
Follow up of #15647 and part of #14510

Use fused type to either compute on 32 or 64 bits depending on the number of bits selected by the user.
The current implementation allow a variable number of bits in [0, 64]. For bits in [0, 32], this PR use `np.uint32` ~and `np.float32`~ types, while values in [33, 64] use `np.uint64` ~and `np.float64`~.

There is the possibility to only allow 3 values for `bits`. `None, 32, 64`. `None` would use 30 bits for backward compatibility. Either here or latter, we can deprecate the value. I would default to 32 bits as most users would not need to go beyond 2^32 points. This will be slightly more performant.

For now I am using `None` which default to 30 bits.

I am hesitant to add a deprecation warning yet. This would mean updating a few areas in our code. `shgo` and `differential_evolution` are using this and it would be a backward incompatibility. We cannot properly show with a warning because `Sobol`'s arguments are not part of the public API of the optimizers. So it would have to go in the release notes maybe during a few releases before we change this. Thoughts?

(I am waiting for this PR to land to update our release notes as this can change things in different directions.)

### CODE REVIEW

**Comment by tupui** [comment]
The dtype seem to be working except that fused types does not work. I have the following declarations:

```c
ctypedef fused uint_32_64:
    cnp.uint32_t
    cnp.uint64_t

ctypedef fused float_32_64:
    cnp.float32_t
    cnp.float64_t
```

If I am using 32 bits, everything works. But if I have 64, then it fails with:

```python
E   ValueError: Buffer dtype mismatch, expected 'npy_uint32' but got 'unsigned long'
```

If I comment the 32 bit declarations and just have:

```c
ctypedef fused uint_32_64:
    cnp.uint64_t

ctypedef fused float_32_64:
    cnp.float64_t
```

Then it works. From my tests, the first type defined in the fused part is used.

@da-woods this is the real PR, sorry for the double ping, we just merged what was needed for this.

I tried to follow your indications (thanks again!). Since the rest is working, I am pretty sure it must be something simple with the way I am calling these. Also if you have more comment in general here, please don't hesitate. I am still a novice in Cython and there is for sure lot of room for improvements.

File: scipy/stats/_sobol.pyx
Comment (da-woods) [inline scipy/stats/_sobol.pyx:128]: Probably as efficient to delegate this copying to Numpy, given that you aren't typing `poly`, `dns_poly` etc.

```suggestion
    poly[:MAXDIM] = dns_poly[:MAXDIM]
    vinit[:MAXDIM, :MAXDEG] = dns_vinit[:MAXDIM, :MAXDEG]
```

File: scipy/stats/_sobol.pyx
Comment (da-woods) [inline scipy/stats/_sobol.pyx:230]: Comment is really on the line `v[d, j] = vinit[d][j]` a little further down (but I can't post one there)

It's usually better to do `vinit[d, i]` (i.e. both indices at once). It avoids creating an intermediate memoryview.

**Review by da-woods — COMMENTED** [review]
> If I am using 32 bits, everything works. But if I have 64, then it fails with:

>> E   ValueError: Buffer dtype mismatch, expected 'npy_uint32' but got 'unsigned long'

Do you know where this error comes from (does it produce a traceback)? I can't see anything obviously wrong, but that may be because I'm looking in the wrong place

File: scipy/stats/_sobol.pyx
Comment (da-woods) [inline scipy/stats/_sobol.pyx:274]: I'm guessing a bit here, but I wonder if `n`, `num_gen`, `scale` are causing the problem. Where they're called from Python they're called with a Python integer for `n`. I'm not sure which int type that'll select, but I think that'll set the type required for `sv` and the array arguments.

Maybe try leaving the arguments untyped (as Python types) as then do `cdef uint_32_64 n = n_py` (where `n_py` is the argument) inside the function.

That will hopefully let Cython select the type based on the type of the array arguments.

File: scipy/stats/_sobol.pyx
Comment (tupui) [reply scipy/stats/_sobol.pyx:128]: Great, didn't know about that and it's faster.

File: scipy/stats/_sobol.pyx
Comment (tupui) [reply scipy/stats/_sobol.pyx:230]: Good catch! A bit faster as well.

File: scipy/stats/_sobol.pyx
Comment (tupui) [reply scipy/stats/_sobol.pyx:274]: That's exactly the issue! I tried to remove the types from all three (and the `nogil` then) and all tests are passing now 😃 

Shall I do two functions then? A `def` wrapping a `cdef`?

**Review by tupui — COMMENTED** [review]
Thank you for the review @da-woods. I will push your two suggestions and try to wrap `_draw` which indeed is making trouble (sorry I should have included since the beginning the traceback. Although it just pointed out to the function without any other message than what I wrote).

If you see anything else in this file, let me know and I will happily fix these.

File: scipy/stats/_sobol.pyx
Comment (da-woods) [reply scipy/stats/_sobol.pyx:271]: A `def` wrapping a `cdef` would be the way to do it. It's only worthwhile creating the extra wrapper layer if you need to call it efficiently from within Cython (otherwise just make the `def`)

Comment (tupui) [reply scipy/stats/_sobol.pyx]: I used a wrapper since it was 10x slower otherwise.

**Comment by tupui** [comment]
@da-woods I finished the Cython parts, could you have another look/review (please approve when changes are good for you, it will help other reviewers)?

File: scipy/stats/_qmc.py
Comment (mdhaber) [inline scipy/stats/_qmc.py:1226]:
```suggestion
        It also conditions the return dtype to be either ``np.float32`` or ``np.float64``.
```
(If that is correct)

Comment (mdhaber) [inline scipy/stats/_qmc.py:1226]:
```suggestion
        It also conditions the return dtype to be either ``np.float32`` or ``np.float64``.
```
(If that is correct)

File: scipy/stats/_qmc.py
Comment (mdhaber) [inline scipy/stats/_qmc.py:1227]: Although with the above, this suggests that it will return a 32-bit array, which I don't think is right.

File: scipy/stats/_qmc.py
Comment (tupui) [reply scipy/stats/_qmc.py:1226]: Totally, clearer indeed!

File: scipy/stats/_qmc.py
Comment (tupui) [reply scipy/stats/_qmc.py:1227]: Yes, I will add something. Shall I constrain the dtype to be 64bits for backward compatibility or do you think it's ok to have a dtype change without notice. As in this case we don't lose precision.

File: scipy/stats/_qmc.py
Comment (mdhaber) [inline scipy/stats/_qmc.py:1719]: Why specify `bits=30` here? If that is the default, why not `None` or just leave it unspecified?

Comment (tupui) [reply scipy/stats/_qmc.py:1720]: I changed this because locally I started to think about using a deprecation warning in case of `None`. So I did not want to raise a warning in such cases (I think)

File: scipy/stats/_qmc.py
Comment (mdhaber) [reply scipy/stats/_qmc.py:1228]: > As in this case we don't lose precision.
The very first points of sequences are different.

```python3
rng = Sobol(1)
x = rng.random()
y = x.astype(np.float32)
x-y  # array([[1.86264515e-08]])
```
Unrelated: I just tried `rng = Sobol(1); x = rng.random(100)` and got the expected warning, but it needs `stacklevel`._

**Comment by da-woods** [comment]
On a quick look through it looks good to me - I can't see any obvious improvements in the Cython code and the initialization stuff looks reasonable.

It is a fairly quick look though, so don't take my approval to mean too much

**Review by da-woods — APPROVED** [review]

**Comment by tupui** [comment]
@mdhaber the CI is green for once!!!

File: scipy/stats/_qmc.py
Comment (tupui) [reply scipy/stats/_qmc.py]: Ah good catch!

Ok I will send an email.

File: scipy/stats/_qmc.py
Comment (tupui) [reply scipy/stats/_qmc.py:1349]: We could do this. But the resolution might be very low. Between 0-1, it does not leave much possibilities.

```python
2**16
# 65536
np.finfo(np.float16)
# finfo(resolution=0.001, min=-6.55040e+04, max=6.55040e+04, dtype=float16)
```

It's more or less the same problem with 32 bits but the limit is higher and will saturate further down the line which makes it less of an issue I think.

```python
2**32
# 4294967296
np.finfo(np.float32)
# finfo(resolution=1e-06, min=-3.4028235e+38, max=3.4028235e+38, dtype=float32)
```

But maybe the output type should be different and not linked to `bits`. I linked both for simplicity for the user.

Comment (tupui) [reply scipy/stats/_qmc.py:1349]: I am actually asking myself if we should not be able to store at least of the possibilities. Like for 32 bits, we should have float64 because there is just a resolution of 1e-6 which is not enough (if my interpretation is correct). I will make some tests and report back. This might simplify the deprecation decision 👍

Comment (tirthasheshpatel) [reply scipy/stats/_qmc.py:1349]: > We could do this. But the resolution might be very low.

Ah, makes sense. Yeah, let's avoid float16 then!

Comment (tirthasheshpatel) [reply scipy/stats/_qmc.py:1349]: > Like for 32 bits, we should have float64 because there is just a resolution of 1e-6 which is not enough (if my interpretation is correct). I will make some tests and report back. This might simplify the deprecation decision

Sounds good. I see returning only float64 has two benefits:

- Simple for the user and doesn't break the backward compatibility.
- Higher precision

So, I think just returning float64 would be better than to switch to float32 for lower bits.

Comment (tupui) [reply scipy/stats/_qmc.py:1349]: Ok so I did some testing. With `bits=32`, if we sample 2^25 then some elements appears twice. Increasing the precision to float64 removes this issue. So for now I will simplify things and just use float64 for the return type.

We could add `dtype` parameter but it would mean that the user would have to know the limits of the type they use with respect to the number of bits they use. I am not sure how to calculate this exactly. It might be confusing anyway.

I will remove the fused type on the return in the Cython part.

**Review by mdhaber — APPROVED** [review]
I don't see any problems. Since we already added variable bits, what did this end up doing? Should there be any tests, or is this just a refactoring now aside from (maybe) some efficiency with integer types? Maybe update the description. @tirthasheshpatel please merge if this looks OK to you.

**Comment by tupui** [comment]
> I don't see any problems. Since we already added variable bits, what did this end up doing? Should there be any tests, or is this just a refactoring now aside from (maybe) some efficiency with integer types? Maybe update the description. @tirthasheshpatel please merge if this looks OK to you.

Thanks Matt. Yes in the end this is changing some internals and the current tests are covering the changes already since we have tests bellow and above 32 bits.

File: scipy/stats/_qmc.py
Comment (tirthasheshpatel) [inline scipy/stats/_qmc.py:1227]:
```suggestion
        ``np.float64`` to prevent points from repeating themselves.
```

File: scipy/stats/_sobol.pyx
Comment (tirthasheshpatel) [inline scipy/stats/_sobol.pyx:41]: This looks cleaner than calling _initialize_direction_numbers in the Python class.

File: scipy/stats/_sobol.pyx
Comment (tirthasheshpatel) [inline scipy/stats/_sobol.pyx:150]: Just a bit concise.

```suggestion
    poly[...] = dns_poly
    vinit[...] = dns_vinit
```

File: scipy/stats/_sobol.pyx
Comment (tirthasheshpatel) [inline scipy/stats/_sobol.pyx:268]: How about: `v[d, :m] = vinit[d, :m]`

**Review by tirthasheshpatel — COMMENTED** [review]
Just a few more comments.

**Review by tirthasheshpatel — APPROVED** [review]
LGTM now! Let's wait for the CI to pass (modulo unrelated azure failures) and then we can merge.

**Comment by tirthasheshpatel** [comment]
Merged! Thanks @tupui @mdhaber @da-woods!

**Pull request merged**
